sbi/inference/posteriors/score_posterior.py

```diff
@@ -0,0 +1,393 @@
+                 posterior.
+             force_update: Whether to re-calculate the MAP when x is unchanged and
+                 have a cached value.
+             log_prob_kwargs: Will be empty for SNLE and SNRE. Will contain
+                 {'norm_posterior': True} for SNPE.
+
+         Returns:
+             The MAP estimate.
+        """
+        return super().map(
+            x,
+            num_iter,
+            num_to_optimize,
+            learning_rate,
+            init_method,
+            num_init_samples,
+            save_best_every,
+            show_progress_bars,
+            force_update,
+        )
+
+    def _calculate_map(
+        self,
+        num_iter: int = 1000,
+        num_to_optimize: int = 100,
+        learning_rate: float = 1e-5,
+        init_method: Union[str, Tensor] = "posterior",
+        num_init_samples: int = 1000,
+        save_best_every: int = 10,
+        show_progress_bars: bool = False,
+    ) -> Tensor:
+        """Calculate the maximum a posteriori (MAP) estimate of the posterior.
+
+        Uses gradient ascent to find the MAP estimate of the posterior. The gradient is
+        calculated using the score estimator.
+
+        Args:
+            num_iter: Number of interations. Defaults to 1000.
+            num_to_optimize : Note used (API), just for interface. Defaults to 100.
+            learning_rate: Learning rate. Defaults to 1e-5.
+            init_method: Initialization of particles. Defaults to "posterior".
+            num_init_samples: Not used (API). Defaults to 1000.
+            save_best_evey: Not used (API). Defaults to 10.
+            show_progress_bars (bool, optional): _description_. Defaults to False.
+
+        Raises:
+            ValueError: Invalid init method
+
+        Returns:
+            Tensor: MAP
+        """
+
+        with torch.no_grad():
+            if init_method == "posterior":
+                inits = self.sample(
+                    (num_init_samples,), show_progress_bars=show_progress_bars
+                )
+            elif init_method == "proposal":
+                inits = self.proposal.sample((num_init_samples,))  # type: ignore
+            elif isinstance(init_method, Tensor):
+                inits = init_method
+            else:
+                raise ValueError
+
+            self.potential_fn_gradient.set_x(self.default_x)
+            gradient_fn = partial(
+                self.potential_fn_gradient,
+                time=torch.tensor([self.score_estimator.T_min]),
+            )
+
+            # Run MAP optimization
+            xs = inits.clone()
+            for _ in range(num_iter):
+                gradient = gradient_fn(xs)
+                xs = xs + learning_rate * gradient
+
+            log_prob = self.log_prob(xs)
+            best_idx = torch.argmax(log_prob)
+            best_theta = xs[best_idx]
+            return best_theta
+
+    def build_freeform_jacobian_transform(
```

Comment:
some documentation needed

Reply:
for context: uses `zuko` to build a Continuous time flow using the score network. Just like it is done for flow matching.